Minor refactoring and improve the performance and reliability of event dots.

ikarosdev opened the pull request:
- Take some of the refactoring work done by @Dirli.
- Increase performance of indicator while scrolling. With no dots in the indicator, the scrolling is smooth between months. If dots are present, the performance takes a hit and is sluggish while moving. This seems to fix that issue.
- Improve reliability of dots showing up. In the case of 3 dots, sometimes the indicator would require a couple openings to show. In the case of 3 dots, oftentimes the indicator would not show 3 dots upon startup. This seems to make event dots appear more reliably for the user.

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -40,6 +41,7 @@ namespace DateTime.Widgets {
         private Gtk.Revealer[] week_labels;
 
         construct {
+            model = Widgets.CalendarModel.get_default ();
```
If we want the model to be static, probably we should create it in a `static construct {}` block

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
+                
+        public void add_event_dots (E.Source source, Gee.Collection<ECal.Component> events) {
+            foreach (var component in events) {
+            
```
Extra whitespace here :)

danrabbit commented on src/Widgets/calendar/GridDay.vala:
```diff
@@ -43,8 +42,7 @@ public class DateTime.Widgets.GridDay : Gtk.EventBox {
     static construct {
-        model = Widgets.CalendarModel.get_default ();
-
+    
```
Extra whitespace here too

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
+            
+                unowned ICal.Component ical = component.get_icalcomponent ();
+                ICal.Time? start_time = ical.get_dtstart ();
+                time_t start_unix = start_time.as_timet ();
```
Since `start_time` is nullable, this probably needs a null check

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
+                var t = new GLib.DateTime.from_unix_utc (start_unix);
+                var d_hash = day_hash (t);
+                if (data.has_key (d_hash)) {
+                    data[d_hash].add_dots (source, component.get_icalcomponent ());
```
You already have `component.get_icalcomponent ()` cached as `ical` right? Probably don't need to get it again

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
         }
+        
+                
+        public void add_event_dots (E.Source source, Gee.Collection<ECal.Component> events) {
```
This seems to only be used internally right? So it can probably be `private`

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
+            }
+        }
+
+        public void remove_event_dots (E.Source source, Gee.Collection<ECal.Component> events) {
```
Same as above, this can probably be `private`

danrabbit commented on src/Widgets/calendar/Grid.vala:
```diff
@@ -260,6 +264,36 @@ namespace DateTime.Widgets {
+                var event_uid = ical.get_uid ();
+
+                ICal.Time? start_time = ical.get_dtstart ();
+                time_t start_unix = start_time.as_timet ();
```
Same as above, probably needs a null check

danrabbit commented on src/Widgets/calendar/GridDay.vala:
```diff
@@ -84,53 +82,49 @@ public class DateTime.Widgets.GridDay : Gtk.EventBox {
-                dot.destroy ();
-                event_dots.remove (event_uid);
-            }
+        var dot = event_grid.get_children ();
+        if (dot.length () > 0) {
+            dot.nth_data (0).destroy ();
```
This seems less efficient and less accurate than using the hashmap as we have it in master. There's no guarantee that the first dot is the correct one to remove, which means you could have dots with colors that don't match what's actually on that day

Dirli replied:
Currently, all your points are depersonalized. Just point with a single class, why should they be personalized? If you try to add, say, 6 events, only 3 will be added. Now that you delete any of the added 3 events, what happens? correctly, the point will be deleted. As a result, you will get 2 points and 5 events.
The code that you reject at least checks whether it is time to delete a point, or only reduce the number of events for this point

ikarosdev replied:
What's a better solution for this? Should I run an iterator, search the list and destroy the element that way?

Dirli replied:
I outlined my opinion above. I hope someone from the developers will voice what logic they laid. I assume that I am wrong, because I do not use this functionality)

ikarosdev replied:
@Dirli and @danrabbit. I think I understand what @danrabbit is saying. I still feel an arraylist is faster than the hashmap implementation. The problem is the dots are color coded for different types of events (I've never noticed this before). So, if we always remove the first dot, that might not be the right one to remove.

ikarosdev replied:
@Dirli I'm not a developer so I'm struggling here. Is there a way to search the arraylist and remove an element from that list based on what event_uid is?

Dirli replied:
Yes, the colors were a surprise to me (while fix issue, I did not come across them). My solution needs to be rewritten to work correctly. Leaving the discussion)

ikarosdev commented:
Thank you for the feedback. I'll clean it up later today when I have some time. I'm not a programmer (just a hobbyist), so I'm learning what's "proper" and not. As for the null checks, does that mean you want a conditional there (i.e. if (!null))?

danrabbit commented:
Definitely a good job cherrypicking then for not considering yourself a programmer!

Right, so if we end up fetching a `null` start time then we won't be able to convert it, hash it, etc because it doesn't exist. So we'll likely want something there that says if `ical.get_dtstart ()` returns `null` then we just stop there and skip to the next component

Dirli commented:
https://valadoc.org/libical-glib/ICal.Component.get_dtstart.html
is an null value returned?

ikarosdev commented:
Alright, so be gentle (seriously, I have no idea what I'm doing). But I added the conditionals you requested. I believe I returned GridDay back to how the function works upstream and I cleaned up any rift-raft whitespace complaints.

ikarosdev commented:
I just realized I borked this. I’ll have to commit a fix.

ikarosdev commented:
I'm going to close this PR. While on the surface this seems to improve reliability and function. This kills certain functionality of the indicator.

danrabbit commented:
Thank you for starting this and the idea! I submitted a branch that should have the same result here: https://github.com/elementary/wingpanel-indicator-datetime/pull/217

I didn't end up including the null checks since as Dirli pointed it, the API doc doesn't seem to indicate that getting start time should be nullable

ikarosdev commented:
@danrabbit Thanks, man. I'm hoping to become a more frequent contributor around these parts.

danrabbit commented:
@ikarosdev that would be great! I appreciate your effort here. Thank you for submitting this PR and getting the conversation going

ikarosdev commented:
@danrabbit Sorry if this isn't the place, but just to clarify (mostly because I've been working on this all day), but I assumed using ArrayList vs HashMap was working "faster," which improved overall performance of the indicator. When I went with the changes @Dirli came up with, I found myself in a world of hurt trying to figure out the right calls (I think I'm using this word correctly) work for ArrayList vs HashMap and I was having to dig through Vala docs. However, the "fix" here is less CalendarModels? In simplest terms, can you explain that?

danrabbit commented:
@ikarosdev right so what’s happening here is we have one grid per month (which honestly this class should be renamed so that’s clear) and inside that grid we have one GridDay per day.

In master, each day is connecting to the add and remove events from the calendar model. So per month that’s typically 30 times we’re running this method that has to figure out if the event is meant for this month and then add it or not.

What this refactor does is instead puts those connections in the month grid and the month grid will figure out which days to add the events to. So we’re doing far less work each time we build a new month grid.
